label_studio/projects/migrations/0028_auto_20241107_1031.py

```diff
@@ -0,0 +1,86 @@
+# Generated by Django 4.2.16 on 2024-11-07 10:31
+
+from django.db import migrations
+import logging
+
+from core.models import AsyncMigrationStatus
+from core.redis import start_job_async_or_sync
+
+logger = logging.getLogger(__name__)
+migration_name = '0028_auto_20241107_1031'
+
+
+def forward_migration(migration_name):
+    from django.db.models import Count, Min
+    from projects.models import ProjectMember
+
+    logger.info(f'Starting async migration {migration_name}')
+    migration = AsyncMigrationStatus.objects.create(
+        name=migration_name,
+        status=AsyncMigrationStatus.STATUS_STARTED,
+    )
+
+    try:
+        # Get projects with duplicates
+        projects_with_duplicates = (
+            ProjectMember.objects
+            .values('project_id', 'user_id')
+            .annotate(entry_count=Count('id'))
+            .filter(entry_count__gt=1)
+            .values_list('project_id', flat=True)
+            .distinct()
+        )
+
+        for project_id in projects_with_duplicates:
+            # Remove duplicates for each project
+            duplicates = (
+                ProjectMember.objects
+                .filter(project_id=project_id)
+                .values('user_id')
+                .annotate(count=Count('id'), min_id=Min('id'))
+                .filter(count__gt=1)
+            )
+            total_deleted = 0
+            for dup in duplicates:
+                user_id = dup['user_id']
+                min_id = dup['min_id']
+                entries_to_delete = (
+                    ProjectMember.objects
+                    .filter(user_id=user_id, project_id=project_id)
+                    .exclude(id=min_id)
+                    .only('id')  # Minimizing fields selected
+                )
+                deleted_count, _ = entries_to_delete.delete()
+                total_deleted += deleted_count
+            logger.info(f'Deleted {total_deleted} duplicate ProjectMember entries for project ID {project_id}.')
+
```

Comment:
maybe we can make as atomic transaction? I will be easy to restart this process if it fails on some project

Reply:
I don't see a big difference with `atomic` because we have only "remove all duplicates for one member" updating query, if it would stop at any moment, we can restart it, it will select again all projects with duplicates and iterate though them again. I some project don't have duplicates anymore, it will not be includes into queryset.